File: csharp/commonTestUtils/ParseKeyValuePair.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace CommonTestUtils
{
    /// <summary>
    /// Parse received line to key value pairs, line example : 
    /// 2016-06-23 00:16:17.601276 from 'Microsoft Windows NT 6.2.9200.0' 'QUAPC' 127.0.0.1 times[1] send message[57] to 10.172.120.118:1982 : tick = 636022377776012761
    /// </summary>
    /// <typeparam name="ValueType"></typeparam>
    [Serializable]
    public abstract class ParseKeyValuePairBase<ValueType, ClassName> : BaseTestUtilLog<ClassName>
    {
        private static long LineCount = 0;

        protected readonly long valueArrayElements;

        protected readonly bool needPrintMessage;

        public ParseKeyValuePairBase(long valueElementCount = 0, bool needPrintMessage = true)
        {
            this.valueArrayElements = valueElementCount;
            this.needPrintMessage = needPrintMessage;
        }

        public virtual void ShowReceivedLine(string line)
        {
            LineCount++;
            if (needPrintMessage)
            {
                Logger.LogInfo($"Received Line[{LineCount}]: {line}");
            }
        }

        public virtual KeyValuePair<string, ValueType> Parse(string line)
        {
            throw new NotImplementedException();
        }

        protected static KeyValuePair<string, int[]> Parse(string line, long elements)
        {
            var match = Regex.Match(line, @"^(?<Key>[\d-]+ [\d:]+)\.(?<Value>\d+)");
            var key = match.Groups["Key"].Value;
            if (string.IsNullOrWhiteSpace(key))
            {
                throw new Exception($"cannot parse key from line : {line}");
            }

            var valueSet = Regex.Matches(match.Groups["Value"].Value, @"\d");
            if (elements == 0)
            {
                elements = valueSet.Count;
            }

            var values = new int[elements];
            var n = Math.Min(valueSet.Count, values.Length);
            for (var k = 0; k < n; k++)
            {
                values[k] = 1; // int.Parse(valueSet[k].Value);
            }

            return new KeyValuePair<string, int[]>(key, values);
        }

        protected void Print(KeyValuePair<string, int[]> kv)
        {
            if (needPrintMessage)
            {
                Logger.LogInfo($"key = {kv.Key} , {TestUtils.ArrayToText("value", kv.Value)}");
            }
        }
    }

    [Serializable]
    public class ParseKeyValueArray : ParseKeyValuePairBase<int[], ParseKeyValueArray>
    {
        public ParseKeyValueArray(long valueElementCount = 0, bool needPrintMessage = true) : base(valueElementCount, needPrintMessage) { }
        public override KeyValuePair<string, int[]> Parse(string line)
        {
            ShowReceivedLine(line);
            var kv = Parse(line, this.valueArrayElements);
            Print(kv);
            return kv;
        }
    }

    [Serializable]
    public class ParseKeyValueUnevenArray : ParseKeyValueArray
    {
        private static Random random = new Random(DateTime.Now.Second);

        public ParseKeyValueUnevenArray(long valueElementCount = 0, bool needPrintMessage = true) : base(valueElementCount, needPrintMessage) { }

        public override KeyValuePair<string, int[]> Parse(string line)
        {
            ShowReceivedLine(line);
            var kv = base.Parse(line);
            var values = kv.Value.ToList();
            int removeCount = random.Next() % (values.Count + 1);
            values.RemoveRange(0, removeCount);
            var pair = new KeyValuePair<string, int[]>(kv.Key, values.ToArray());
            Print(pair);
            return pair;
        }
    }

    [Serializable]
    public class ParseKeyValue : ParseKeyValuePairBase<int, ParseKeyValue>
    {
        public ParseKeyValue(long valueArrayElements = 0, bool needPrintMessage = true) : base(valueArrayElements, needPrintMessage) { }

        public override KeyValuePair<string, int> Parse(string line)
        {
            ShowReceivedLine(line);
            var kv = Parse(line, 1);
            Print(kv);
            return new KeyValuePair<string, int>(kv.Key, kv.Value[0]);
        }
    }
}

File: scala/kafkaStreamTest/src/main/scala/lzTest/ParseKeyValuePair.scala
package lzTest

import java.util.Calendar

import org.apache.commons.lang.NotImplementedException

import scala.util.Random
import scala.util.matching.Regex

/**
  * Created by qualiu on 6/22/2016.
  */

abstract class ParseKeyValuePairBase[ValueType](val valueArrayElements: Int = 0, val needPrintMessage: Boolean = true) extends LogBase {
  //  def log(message: String): Unit = {
  //    println(s"${TestUtil.NowMilli} : ${this.getClass.getName} $message")
  //  }

  def parse(line: String): Tuple2[String, ValueType] = {
    throw new NotImplementedException
  }

  def printTuple(kv: Tuple2[String, Array[Int]]): Unit = {
    if (needPrintMessage) {
      log(s"key = ${kv._1} , ${TestUtil.ArrayToText("value", kv._2)}")
    }
  }

  // 2016-06-23 00:16:17.601276 from 'Microsoft Windows NT 6.2.9200.0' 'QUAPC' 127.0.0.1 times[1] send message[57] to 10.172.120.118:1982 : tick = 636022377776012761
  def parseToArray(line: String, elements: Int): Tuple2[String, Array[Int]] = {
    log(s"received line = ${line}")
    // val regLine = new Regex("""^.*tick\s*=\s*(?<Key>\d{9})(?<Value>\d+)""", "Key", "Value")
    val regLine = new Regex("""^([\d-]+ [\d:]+)\.(\d+)""", "Key", "Value")
    val matched = regLine.findFirstMatchIn(line).get
    val key = matched.group("Key")
    val valueSet = """\d""".r.findAllMatchIn(matched.group("Value")).toArray
    val values = new Array[Int](math.max(1, elements))
    for (k <- 0 until math.min(valueSet.length, values.length)) {
      values(k) = 1 // valueSet(k).group(0).toInt
    }
    new Tuple2[String, Array[Int]](key, values)
  }
}

class ParseKeyValueArray(valueArrayElements: Int = 0, needPrintMessage: Boolean = true)
  extends ParseKeyValuePairBase[Array[Int]](valueArrayElements, needPrintMessage) {
  override def parse(line: String): Tuple2[String, Array[Int]] = {
    val kv = parseToArray(line, valueArrayElements)
    printTuple(kv)
    kv
  }
}

class ParseKeyValueUnevenArray(valueArrayElements: Int = 0, needPrintMessage: Boolean = true)
  extends ParseKeyValuePairBase[Array[Int]](valueArrayElements, needPrintMessage) {
  val random = new Random(System.currentTimeMillis) //new Random(Calendar.getInstance().getTimeInMillis)

  override def parse(line: String): Tuple2[String, Array[Int]] = {
    val kv = parseToArray(line, valueArrayElements)
    val values = kv._2.toList
    val removeCount = random.nextLong() % (values.length + 1)
    //values.remove(0, removeCount)
    val newArray = values.takeRight((values.length - removeCount).toInt)
    var pair = new Tuple2[String, Array[Int]](kv._1, values.toArray)
    printTuple(pair)
    pair
  }
}

class ParseKeyValue(valueArrayElements: Int = 0, needPrintMessage: Boolean = true)
  extends ParseKeyValuePairBase[Int](valueArrayElements, needPrintMessage) {
  override def parse(line: String): Tuple2[String, Int] = {
    var kv = parseToArray(line, 1)
    printTuple(kv)
    new Tuple2[String, Int](kv._1, kv._2(0))
  }
}
